smithy-typescript-ssdk-libs/server-common/src/httpbinding/mux.ts

```diff
@@ -0,0 +1,146 @@
+ *
+ *   http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+import { HttpRequest } from "@aws-sdk/types";
+import { Mux, ServiceCoordinate } from "..";
+
+export interface PathLiteralSegment {
+    type: 'path_literal'
+    value: string
+}
+
+export interface PathLabelSegment {
+    type: 'path'
+}
+
+export interface GreedySegment {
+    type: 'greedy'
+}
+
+export interface QueryLiteralSegment {
+    type: 'query_literal'
+    key: string
+    value?: string
+}
+
+export interface QuerySegment {
+    type: 'query'
+    key: string
+}
+export class UriSpec<S extends string, O extends String> {
+    private readonly method: string;
+    private readonly pathSegments: (PathLiteralSegment | PathLabelSegment | GreedySegment)[];
+    private readonly querySegments: (QueryLiteralSegment | QuerySegment)[];
+    readonly rank: number;
+    readonly target: ServiceCoordinate<S, O>;
+
+    constructor(method: string,
+                pathSegments: (PathLiteralSegment | PathLabelSegment | GreedySegment)[],
+                querySegments: (QueryLiteralSegment | QuerySegment)[],
+                target: ServiceCoordinate<S, O>) {
+        this.method = method;
+        this.pathSegments = pathSegments;
+        this.querySegments = querySegments;
+        this.rank = this.pathSegments.length + this.querySegments.length;
+        this.target = target;
+    }
+
+    private matchesSegment(requestSegment: string,
+        segment: PathLiteralSegment | PathLabelSegment | GreedySegment): boolean {
+        if (segment.type === 'path_literal' && requestSegment !== segment.value) {
+            return false;
+        }
+        return true;
+    }
+
+    match(req: HttpRequest) : boolean {
+        if (req.method !== this.method) {
+            return false;
+        }
+
+        let requestPathSegments = req.path.split("/").filter(s => s.length > 0);
+
+        let requestPathIdx = 0;
+        path_loop: for (let i = 0; i < this.pathSegments.length; i++) {
+            if (requestPathIdx === requestPathSegments.length) {
+                // there are more pathSegments but we have reached the end of the requestPath
+                return false;
+            }
+            const pathSegment = this.pathSegments[i];
+            if (pathSegment.type === 'path_literal' && pathSegment.value !== requestPathSegments[requestPathIdx]) {
+                return false;
+            }
+            if (pathSegment.type === 'greedy') {
+                if (i === this.pathSegments.length - 1) {
+                    // greedy label at the end of pathSegments swallows the remaining path segments
+                    requestPathIdx = requestPathSegments.length;
+                    break path_loop;
```

Comment:
Is`path_loop` here and line 76 necessary?

Reply:
Nah, it's a vestige of when this code was even harder to follow (hard to believe, I know). I kept it in case it made the flow of the break more understandable since there's a lot of nesting, not because the break was ambiguous.